lib/network/connection_manager.go

```diff
@@ -1,268 +1,19 @@
-			continue
-		}
-		connected = append(connected, address)
-	}
-
-	return connected
-}
-
-// Returns:
-//   A list of all validators, including self
-func (c *ConnectionManager) AllValidators() []string {
-	var validators []string
-	for address := range c.validators {
-		validators = append(validators, address)
-	}
-	return append(validators, c.localNode.Address())
-}
-
-//
-// Returns:
-//   the number of validators which are currently connected
-//
-func (c *ConnectionManager) CountConnected() int {
-	c.RLock()
-	defer c.RUnlock()
-	return c.countConnectedUnlocked()
-}
-
-func (c *ConnectionManager) countConnectedUnlocked() int {
-	var count int
-	for _, isConnected := range c.connected {
-		if isConnected {
-			count += 1
-		}
-	}
-	return count
-}
-
-func (c *ConnectionManager) connectingValidator(v *node.Validator) {
-	ticker := time.NewTicker(time.Second * 1)
-	for _ = range ticker.C {
-		err := c.connectValidator(v)
-
-		if c.setConnected(v, err == nil) {
-			if err == nil {
-				c.log.Debug("validator is connected", "validator", v)
-			} else {
-				c.log.Debug("validator is disconnected", "validator", v, "error", err)
-			}
-		}
-	}
-
-	return
-}
-
-func (c *ConnectionManager) connectValidator(v *node.Validator) (err error) {
-	client := c.GetConnection(v.Address())
-
-	var b []byte
-	b, err = client.Connect(c.localNode)
-	if err != nil {
-		return
-	}
-
-	// load and check validator info; addresses are same?
-	var validator *node.Validator
-	validator, err = node.NewValidatorFromString(b)
-	if err != nil {
-		return
-	}
-	if v.Address() != validator.Address() {
-		err = errors.New("address is mismatch")
-		return
-	}
-
-	return
-}
-
-func (c *ConnectionManager) ConnectionWatcher(t Network, conn net.Conn, state http.ConnState) {
-	return
-}
-
-func (c *ConnectionManager) Broadcast(message common.Message) {
-	c.broadcaster.Broadcast(
-		message,
-		func(v string, err error) {
-			c.log.Error("failed to SendBallot", "error", err, "validator", v)
-		},
-	)
-}
-
-type simpleBroadcaster struct {
-	cm *ConnectionManager
-}
-
-func NewSimpleBroadcaster(c *ConnectionManager) *simpleBroadcaster {
-	if c == nil {
-		panic("ConnectionManager is nil")
-	}
-	p := &simpleBroadcaster{
-		cm: c,
-	}
-	return p
-}
-
-func (b *simpleBroadcaster) Broadcast(message common.Message, onError func(string, error)) {
-	b.cm.RLock()
-	defer b.cm.RUnlock()
-	for addr, connected := range b.cm.connected {
-		if connected {
-			go func(v *node.Validator) {
-				client := b.cm.GetConnection(v.Address())
-
-				var err error
-				if message.GetType() == common.BallotMessage {
-					_, err = client.SendBallot(message)
-				} else if message.GetType() == string(common.TransactionMessage) {
-					_, err = client.SendMessage(message)
-				} else {
-					panic("invalid message")
-				}
-
-				if err != nil {
-					onError(v.Address(), err)
-				}
-			}(b.cm.validators[addr])
-		}
-	}
-	return
+type ConnectionManager interface {
+	CalculateProposer(uint64, uint64) string
```

Comment:
I feel like that `CalculateProposer` is not fit in `ConnectionManager`.. Isn't it? :->

Reply:
Buy `ConnectionManager` interface is good! :+1:

Reply:
Then where is good position for `CalculateProposer` you think? :)

Reply:
I agree. `CalculateProposer` is consensus-related, not network-related.